util/auto_tune_compressor.cc

```diff
@@ -131,151 +123,235 @@ std::unique_ptr<Compressor> AutoSkipCompressorManager::GetCompressorForSST(
-                                          ManagedWorkingArea* wa) {
+Status AutoTuneCompressor::CompressBlock(Slice uncompressed_data,
+                                         std::string* compressed_output,
+                                         CompressionType* out_compression_type,
+                                         ManagedWorkingArea* wa) {
   // Check if the managed working area is provided or owned by this object.
-  // If not, bypass compressor logic since the working area lacks a predictor
-  if (allcompressors_.size() == 0) {
-    return Status::NotSupported("No compression type supported");
-  }
+  // If not, bypass compressor logic since the working area lacks a predictor.
   if (wa == nullptr || wa->owner() != this) {
-    // highest compression level of Zstd
-    size_t choosen_compression_type = 6;
-    size_t compression_level_ptr = 2;
-    return allcompressors_[choosen_compression_type][compression_level_ptr]
-        ->CompressBlock(uncompressed_data, compressed_output,
-                        out_compression_type, wa);
+    return default_compressor_->CompressBlock(
+        uncompressed_data, compressed_output, out_compression_type, wa);
+  }
+
+  if (io_goal_ == nullptr || cpu_budget_ == nullptr) {
+    return Status::InvalidArgument("IOGoal or CPUBudget is not set");
+  } else if (compressors_.size() == 0) {
+    return Status::NotSupported(
+        "No compression alorithm that AutoTune compressor can use are "
+        "supported (SNAPPY, LZ4, LZ4HC, ZSTD)");
   }
   auto local_wa = static_cast<CostAwareWorkingArea*>(wa->get());
-  std::pair<size_t, size_t> choosen_index(6, 2);
-  size_t choosen_compression_type = choosen_index.first;
-  size_t compresion_level_ptr = choosen_index.second;
-  return CompressBlockAndRecord(choosen_compression_type, compresion_level_ptr,
-                                uncompressed_data, compressed_output,
-                                out_compression_type, local_wa);
+  bool exploration =
+      Random::GetTLSInstance()->PercentTrue(kExplorationPercentage);
+  TEST_SYNC_POINT_CALLBACK(
+      "AutoTuneCompressorWrapper::CompressBlock::exploitOrExplore",
+      &exploration);
+  if (exploration) {
+    size_t choosen_index = Random::GetTLSInstance()->Uniform(
+        static_cast<int>(compressors_.size()));
+
+    return CompressBlockAndRecord(choosen_index, uncompressed_data,
+                                  compressed_output, out_compression_type,
+                                  local_wa);
+  } else {
+    auto chosen_compressor = SelectCompressionBasedOnIOGoalCPUBudget(local_wa);
+    TEST_SYNC_POINT_CALLBACK(
+        "AutoTuneCompressorWrapper::CompressBlock::GetSelection",
+        &chosen_compressor);
+    // Check if the chosen compression type and level are available
+    // if not, skip the compression
+    if (chosen_compressor >= compressors_.size()) {
+      *out_compression_type = kNoCompression;
```

Comment:
I would suggest making the check for "valid choice" into the  SelectCompressionBasedOnIOGoalCPUBudget.